service/history/hsm/tree.go

```diff
@@ -0,0 +1,366 @@
+	serialized, err := def.Serialize(data)
+	if err != nil {
+		return nil, err
+	}
+	node := &Node{
+		Key:        key,
+		Parent:     n,
+		definition: def,
+		registry:   n.registry,
+		persistence: &persistencespb.StateMachineNode{
+			Children: make(map[int32]*persistencespb.StateMachineMap),
+			Data:     serialized,
+		},
+		cache: &cachedMachine{
+			dataLoaded: true,
+			data:       data,
+			children:   make(map[Key]*Node),
+		},
+	}
+	n.cache.children[key] = node
+	children, ok := n.persistence.Children[key.Type]
+	if !ok {
+		children = &persistencespb.StateMachineMap{MachinesById: make(map[string]*persistencespb.StateMachineNode)}
+		n.persistence.Children[key.Type] = children
+	}
+	children.MachinesById[key.ID] = node.persistence
+	return node, nil
+}
+
+// MachineData deserializes the persistent state machine's data, casts it to type T, and returns it.
+// Returns an error when deserialization or casting fails.
+func MachineData[T any](n *Node) (T, error) {
+	var t T
+	if n.cache.dataLoaded {
+		if t, ok := n.cache.data.(T); ok {
+			return t, nil
+		}
+		return t, ErrIncompatibleType
+	}
+	a, err := n.definition.Deserialize(n.persistence.Data)
+	if err != nil {
+		return t, err
+	}
+	n.cache.data = a
+	n.cache.dataLoaded = true
+
+	if t, ok := a.(T); ok {
+		return t, nil
+	}
+	return t, ErrIncompatibleType
+}
+
+// TransitionCount returns the transition count for the state machine contained in this node.
+func (n *Node) TransitionCount() int64 {
+	return n.persistence.TransitionCount
+}
+
+// MachineTransition runs the given transitionFn on a machine's data for the given key.
+// It updates the state machine's metadata and marks the entry as dirty in the node's cache.
+// If the transition fails, the changes are rolled back and no state is mutated.
+func MachineTransition[T any](n *Node, transitionFn func(T) (TransitionOutput, error)) (retErr error) {
+	data, err := MachineData[T](n)
+	if err != nil {
+		return err
+	}
+	n.persistence.TransitionCount++
+	// Rollback on error
+	defer func() {
+		if retErr != nil {
+			n.persistence.TransitionCount--
+			// Force reloading data.
+			n.cache.dataLoaded = false
+		}
+	}()
+	output, err := transitionFn(data)
+	if err != nil {
+		return err
+	}
+	serialized, err := n.definition.Serialize(data)
+	if err != nil {
+		return err
+	}
+	n.persistence.Data = serialized
+	n.cache.outputs = append(n.cache.outputs, output)
+	return nil
+}
+
+// A Collection of similarly typed sibling state machines.
+type Collection[T any] struct {
+	// The type of machines stored in this collection.
+	Type int32
+	node *Node
+}
+
+// NewCollection creates a new [Collection].
+func NewCollection[T any](node *Node, stateMachineType int32) Collection[T] {
+	return Collection[T]{
+		Type: stateMachineType,
+		node: node,
+	}
+}
+
+// Node gets an [Node] for a given state machine ID.
+func (c Collection[T]) Node(stateMachineID string) (*Node, error) {
+	return c.node.Child([]Key{{Type: c.Type, ID: stateMachineID}})
```

Comment:
so a collection contains all state machines of a certain type that are the **_immediate_** children of a given node?

Reply:
> // A Collection of similarly typed sibling state machines.

Does that answer your question?